module/VuFind/src/VuFind/View/Helper/Root/ProxyUrl.php

```diff
@@ -66,10 +80,67 @@ public function __construct($config = null)
      */
     public function __invoke($url)
     {
-        $usePrefix = !isset($this->config->EZproxy->prefixLinks)
-            || $this->config->EZproxy->prefixLinks;
+        $useWebService = $this->config->EZproxy->prefixLinksWebServiceUrl ?? false;
+        if ($useWebService) {
+            $usePrefix = $this->checkUrl($url) ?? $this->checkConfig();
+        } else {
+            $usePrefix = $this->checkConfig();
+        }
+
         return ($usePrefix && isset($this->config->EZproxy->host))
             ? $this->config->EZproxy->host . '/login?qurl=' . urlencode($url)
             : $url;
     }
+
+    /**
+     * Return the configured prefixLinks setting.
+     *
+     * @return bool The configured setting, or the default
+     */
+    private function checkConfig()
+    {
+        return $this->config->EZproxy->prefixLinks ?? true;
+    }
+
+    /**
+     * Check whether the given URL requires the proxy prefix.  Cache the repsonse.
+     *
+     * @param string $url The raw URL to check
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function checkUrl($url)
+    {
+        $domain = parse_url($url, PHP_URL_SCHEME)
+         . '://'
+         . parse_url($url, PHP_URL_HOST);
+        $usePrefix = $this->getCachedData("proxyUrl-domainToUsePrefix-$domain");
+        if (null === $usePrefix) {
+            $usePrefix = $this->queryWebService($domain);
+            $this->putCachedData("proxyUrl-domainToUsePrefix-$domain", $usePrefix);
+        }
+        return $usePrefix;
+    }
+
+    /**
+     * Query the web service on whether to prefix URLs to a given domain.
+     *
+     * @param $domain The domain
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function queryWebService($domain)
+    {
+        $prefixLinksWebServiceUrl = $this->config->EZproxy->prefixLinksWebServiceUrl;
+        $queryUrl = $prefixLinksWebServiceUrl . '?url=' . $domain;
+        try {
+            $client  = $this->httpService->createClient($queryUrl);
+            $response = $client->send();
+            $responseData = $response->getContent();
+        } catch (Exception $ex) {
+            $this->logError('Exception during EZproxy web service request: ' . $ex->getMessage());
+            return null;
```

Comment:
When the service is unavailable, should we fallback to proxying the URL? This would keep the same functionality that currently exists in the event the service is unavailable for whatever reason.

```suggestion
            return TRUE;
```

Reply:
Ah, I see. If `NULL` is returned, `checkConfig` will use the default behavior 👍